domain/relation/state/relation.go

```diff
@@ -481,6 +483,389 @@ WHERE  relation_id = $getRelation.relation_id
+// scope. Under the following conditions, a subordinate unit cannot enter a
+// relation that is not with its principle application:
+//   - The relation has scope container.
+//   - The unit entering scope is a subordinate.
+//   - The other application is not a subordinate.
+func (st *State) checkSubordinateUnitCanEnterScope(
+	ctx context.Context,
+	tx *sqlair.TX,
+	relUUID corerelation.UUID,
+	unitUUID unit.UUID,
+	otherApplication application.ID,
+) error {
+	// Check that the relation is a container scoped relation.
+	scope, err := st.getRelationScope(ctx, tx, relUUID)
+	if err != nil {
+		return errors.Errorf("getting relation scope: %w", err)
+	}
+	if scope != charm.ScopeContainer {
+		return nil
+	}
+
+	// Check that the other application in the relation is not a subordinate
+	// application, if it is, we have a relation between two subordinates, which
+	// is OK.
+	if subordinate, err := st.isSubordinate(ctx, tx, otherApplication); err != nil {
+		return errors.Errorf("checking if application is subordinate: %w", err)
+	} else if !subordinate {
+		return nil
+	}
+
+	// Check that the principle application of the unit is the other application
+	// in the relation the unit is trying to enter scope in.
+	principleAppID, err := st.getPrincipalApplicationOfUnit(ctx, tx, unitUUID)
+	if err != nil {
+		return errors.Errorf("getting principal application of unit: %w", err)
+	}
+	if principleAppID != otherApplication {
+		return errors.Errorf(
+			"unit cannot enter scope: principle application not in relation: %w",
+			relationerrors.PotentialRelationUnitNotValid,
+		)
+	}
+
+	return nil
+}
+
+// getRelationScope returns the scope of the given relation.
+func (st *State) getRelationScope(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid corerelation.UUID,
+) (charm.RelationScope, error) {
+	type getScope struct {
+		Scope charm.RelationScope `db:"scope"`
+	}
+	relUUID := relationUUID{
+		UUID: uuid.String(),
+	}
+	var scope getScope
+	getScopeStmt, err := st.Prepare(`
+SELECT &getScope.*
+FROM   v_relation_endpoint
+WHERE  relation_uuid = $relationUUID.uuid
+`, relUUID, scope)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	err = tx.Query(ctx, getScopeStmt, relUUID).Get(&scope)
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+	return scope.Scope, nil
+}
+
+// isSubordinate returns true if the application is a subordinate application.
+func (st *State) isSubordinate(
+	ctx context.Context,
+	tx *sqlair.TX,
+	applicationUUID application.ID,
+) (bool, error) {
+	type getSubordinate struct {
+		ApplicationUUID application.ID `db:"application_uuid"`
+		Subordinate     bool           `db:"subordinate"`
+	}
+
+	subordinate := getSubordinate{
+		ApplicationUUID: applicationUUID,
+	}
+
+	stmt, err := st.Prepare(`
+SELECT subordinate AS &getSubordinate.subordinate
+FROM   application a
+JOIN   charm c ON a.charm_uuid = c.uuid
+JOIN   charm_metadata cm ON c.uuid = cm.charm_uuid
+WHERE  a.uuid = $getSubordinate.application_uuid
```

Comment:
Be kind to our future selves. Start with the table which holds the data, and be specific where the subordinate comes from.
```suggestion
SELECT cm.subordinate AS &getSubordinate.subordinate
FROM    charm_metadata cm 
JOIN      charm c ON cm.charm_uuid = c.uuid
JOIN      application a  ON c.uuid = a.charm_uuid
WHERE  a.uuid = $getSubordinate.application_uuid
```